api/analyzedphenotypes.datafile.api.inc

```diff
@@ -0,0 +1,175 @@
+  $extensions = array(
+    'file_upload'   => array(
+      'tsv'  => 'Tab Separated Values',
+      'txt'  => 'Text File',
+    ),
+    'file_download' => array(
+      'tsv'  => '.TSV - Tab Separated Values',
+      'csv'  => '.CSV - Comman Separated Values',
+      'xlsx' => '.XLSX - Microsoft Excel Spreadsheet File',
+    )
+  );
+
+  return ($set) ? $extensions[$set] : $extensions;
+}
+
+/**
+ * Define file destination path for file upload and trait image.
+ * 
+ * @return 
+ *   Associative array where the key is directory type and value as directory path.
+ */
+function ap_construct_datafile_destinationpath() {
+  $drupal_public = 'public://';
+  $ap_root_dir = 'tripal/analyzedphenotypes/';
+  
+  // Add directory here. 
+  $directory = array(
+    'spreadsheet',
+    'traitpictures'
+  );
+  
+  // Constuct directories.
+  $file_dir = array();
+  
+  foreach($directory as $dir) {
+    $file_dir[$dir] = $drupal_public . $ap_root_dir . $dir; 
+  }
+  
+  return $file_dir;
+}
+
+///////
+
+
+/**
+ * CONTROL/SWITCHBOARD.
+ * Pass a request to a corresponding function.
+ * @TODO: leave out when api reorganize/update is complete and use functions below.
+ */
+function analyzedphenotypes_datafileprop($property = null, $parameter = null) {
+  if (is_null($property)) {
+    // Default.
+    $property = 'default';
+  }
+
+  // Process request.
+  // Default to all projects.
+  $command = array(
+    'default' => array(
+      'func'  => 'ap_define_datafile_extensions',
+      'param' => 'file_upload',
+    ),
+    'download_type' => array(
+      'func'  => 'ap_define_datafile_extensions',
+      'param' => 'file_download',
+    ),
+    'get_contents' => array(
+      'func'  => 'ap_get_datafile_contents',
+      'param' => $parameter['data_file'],
+    ),
+    'get_columns' => array(
+      'func'  => 'ap_get_datafile_columns',
+      'param' => $parameter['data_file'],
+    ),
+    'make_permanent' => array(
+      'func'  => 'ap_make_datafile_permanent',
+      'param' => $parameter['data_file'],
+    ),
+  );
+
+  return call_user_func($command[$property]['func'], $command[$property]['param']);
+}
+
+
+///////
+
+
+/**
+ * FUNCTION:
+ * Read data file and return file data contents.
+ *
+ * @param $datafile
+ *   Drupal file object.
+ *
+ * @return
+ *   File contents.
+ */
+function ap_get_datafile_contents($datafile) {
+  $file = drupal_realpath($datafile->uri);
+  if (file_exists($file)) {
+    $datafile_contents = file_get_contents($source);
+
+    if ($datafile_contents) {
+      $contents = $datafile_contents;
+    }
+  }
+
+  return ($contents) ? $contents : null;
+}
+
+/**
+ * FUNCTION:
+ * Read data file and return column headers which is the first row of file contents.
+ * Dependencies: ap_get_datafile_columns().
+ *
+ * @param $datafile
+ *   Drupal file object.
+ *
+ * @return
+ *   An array, containing column headers.
+ */
+function ap_get_datafile_columns($datafile) {
+  $contents = ap_get_datafile_contents();
```

Comment:
Do Not Read the whole file into memory just to grab the header! Instead, open the file for reading, read the first line and close it. Biological data files can end up being huge and this would cause serious issues with the system.

Reply:
Underestimated things knowing we have a fix number of columns and rows by the few hundreds. Seeing data in rawphenotypes... Revised and this function does not exist.

I have localized reading of file to where it is required and with this method column headers can be readily accessed, this function too does not exists.

The remaining functions were move to general api/helper functions and data file api file does not exist also.